BLE: Introduce ChainableEventHandler and subclasses

AGlass0fMilk opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

~Requires #13727 to be merged first~
Merged #13727 since this PR is closely tied to it.

Introduces a `ChainableEventHandler` base class that is essentially a singly-linked-list of EventHandlers along with two subclasses:

- `ChainableGattServerEventHandler` that enables chaining together `GattServer::EventHandler`s
- `ChainableGapEventHandler` that enables chaining together `Gap::EventHandlers`s

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

The ChainableGattServerEventHandler class allows an application to `register` separate event handlers (eg: for different services that need to handle `GattServer` events) and then set the global `GattServer::setEventHandler` to the instance of `ChainableGattServerEventHandler` with all registered `GattServer::EventHandler`s.

The ChainableGapEventHandler accomplishes that same as above for `Gap:EventHandler` implementations.

Common functionality has been split off into `ChainableEventHandler`.

See #13728 for discussion around this implementation.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
None
----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@pan- @paul-szczepanek-arm The code compiles but my template-class-function-pointer kung fu isn't flawless. Please look over my implementation and point out anything that looks suspect.

I will test this with an example in the near future.

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

AGlass0fMilk committed:
Update GattServer Callback API

Update parameters passed to onDataSent, onUpdatesEnabled/Disabled, and onConfirmationReceived callbacks.

Deprecate single-callback-registering functions for event handling in lieu of the new EventHandler-based API.

Introduce new GattServer::EventHandler callback functions to replace the deprecated versions.

AGlass0fMilk committed:
Introduce ChainableEventHandler and ChainableGattServerEventHandler classes.

Common functionality has been split off into a generic ChainableEventHandler for use by other EventHandler implementations. The ChainableEventHandler is essentially singly-linked list that propagates callbacks to all objects in the list.

The ChainableGattServerEventHandler enables chaining together GattServer::EventHandlers. An application can register separate event handlers (eg: for different services that need to handle GattServer events) and then set the global GattServer::setEventHandler to the instance of ChainableGattServerEventHandler with all registered GattServer::EventHandlers.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/include/ble/gatt/ChainableGattServerEventHandler.h:
```diff
@@ -0,0 +1,211 @@
+     * @note the returned node pointer should be retained if
+     * the application ever wishes to deregister the specific EventHandler
+     */
+    void* registerEventHandler(ble::GattServer::EventHandler* event_handler) {
```
that's just my personal opinion because I like simple names but addEventHandler and removeEventHandler come more naturally to my mind as they suggest that there are multiple slots

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/include/ble/gatt/ChainableGattServerEventHandler.h:
```diff
@@ -0,0 +1,211 @@
+     *
+     * @param[in] node Pointer to node to deregister
+     */
+    void deregisterEventHandler(void* node) {
```
you have to store an array of nodes you registered yourself to be able to remove them, don't think that's necessary the event handler is already a pointer, we could just pass that to remove it

AGlass0fMilk replied:
True, I have changed this so you just pass in the original event handler pointer.

AGlass0fMilk committed:
Introduce ChainableGapEventHandler subclass of ChainableEventHandler

The ChainableGapEventHandler enables chaining together Gap::EventHandlers so separate parts of an application can be notified of Gap events as needed. The application can register multiple Gap::EventHandler objects to one ChainableGapEventHandler and then set the global Gap::EventHandler to the ChainableGapEventHandler. All registered EventHandlers will then be called when a Gap Event occurs.

pan- commented on connectivity/FEATURE_BLE/include/ble/common/ChainableEventHandler.h:
```diff
@@ -0,0 +1,125 @@
+            (it->eh->*fn)(args...);
+            it = it->next;
+        }
+    }
```
```suggestion
    template<typename... FnArgs, typename... Args>
    void execute_on_all(void (T::*fn)(FnArgs...), Args&&... args) {
        auto it = head;
        while(it) {
            // we do not use std::forward, args have to remain lvalues
            // as they are passed to multiple handlers
            (it->eh->*fn)(args...);
            it = it->next;
        }
    }
```

pan- commented on connectivity/FEATURE_BLE/include/ble/common/ChainableEventHandler.h:
```diff
@@ -0,0 +1,118 @@
+     */
+    void addEventHandler(T* event_handler) {
+        auto eh = new (std::nothrow) node_t { event_handler, nullptr };
+        if(!eh) { return; }
```
I'd like this error to be reported to the user.

pan- commented on connectivity/FEATURE_BLE/include/ble/common/ChainableEventHandler.h:
```diff
@@ -0,0 +1,118 @@
+
+    ChainableEventHandler() { }
+
+    virtual ~ChainableEventHandler() {
```
Does this have to be `virtual` is it expected to do a polymorph `delete` through a `ChainableEventHandler` pointer.

pan- commented on connectivity/FEATURE_BLE/include/ble/gap/ChainableGapEventHandler.h:
```diff
@@ -0,0 +1,274 @@
+     *
+     * @see AdvertisingParameters::setScanRequestNotification().
+     */
+    void onScanRequestReceived(const ble::ScanRequestEvent &event) override {
```
For documentation of the event handler, I'd just reference the parent's documentation so we don't have duplicates that requires update. 

You can either remove the command as the doxygen flag `INHERIT_DOCS` is set to `YES` ([ref](https://www.doxygen.nl/manual/config.html#cfg_inherit_docs)) or  use [`@copydoc`](https://www.doxygen.nl/manual/commands.html#cmdcopydoc)

pan- reviewed (dismissed):
Great change, I left few comments for improvment.

pan- commented on connectivity/FEATURE_BLE/include/ble/gap/ChainableGapEventHandler.h:
```diff
@@ -83,6 +83,7 @@ class ChainableGapEventHandler : public ble::Gap::EventHandler,
     }
 
     void onDisconnectionComplete(const ble::DisconnectionCompleteEvent &event) override {
+       execute_on_all(&ble::Gap::EventHandler::onDisconnectionComplete, event);
```
Thanks, I forgot to comment for that one 😅 .

pan- reviewed (dismissed):
LGTM

pan- reviewed (dismissed):
LGTM
